src/gpu_common.c

```diff
@@ -458,43 +458,60 @@ void gpu_check_temp(void)
 {
 #if HAVE_LIBDL
 	static int warned;
-	int i;
+	int i, hot_gpu = 0;
 
 	if (gpu_temp_limit < 0)
 		return;
 
-	for (i = 0; i < MAX_GPU_DEVICES && engaged_devices[i] != DEV_LIST_END; i++)
-	if (dev_get_temp[engaged_devices[i]]) {
-		int fan, temp, util, cl, ml;
-		int dev = engaged_devices[i];
-
-		dev_get_temp[dev](temp_dev_id[dev], &temp, &fan, &util, &cl, &ml);
-
-		if (temp > 125 || temp < 10) {
-			if (!warned++) {
-				log_event("Device %d probably invalid temp reading (%d%sC).",
-				          dev + 1, temp, gpu_degree_sign);
-				fprintf(stderr,
-				        "Device %d probably invalid temp reading (%d%sC).\n",
-				        dev + 1, temp, gpu_degree_sign);
+	do {
+		for (i = 0; i < MAX_GPU_DEVICES && engaged_devices[i] != DEV_LIST_END; i++)
+		if (dev_get_temp[engaged_devices[i]]) {
+			int fan, temp, util, cl, ml;
+			int dev = engaged_devices[i];
+
+			dev_get_temp[dev](temp_dev_id[dev], &temp, &fan, &util, &cl, &ml);
+
+			if (temp > 125 || temp < 10) {
+				if (!warned++) {
+					log_event("Device %d probably invalid temp reading (%d%sC).",
+					          dev + 1, temp, gpu_degree_sign);
+					fprintf(stderr,
+					        "Device %d probably invalid temp reading (%d%sC).\n",
+					        dev + 1, temp, gpu_degree_sign);
+				}
+				return;
 			}
-			return;
-		}
 
-		if (temp >= gpu_temp_limit) {
-			char s_fan[16] = "n/a";
-			if (fan >= 0)
-				sprintf(s_fan, "%u%%", fan);
-			if (!event_abort) {
-				log_event("Device %d overheat (%d%sC, fan %s), aborting job.",
-				          dev + 1, temp, gpu_degree_sign, s_fan);
-				fprintf(stderr,
-				        "Device %d overheat (%d%sC, fan %s), aborting job.\n",
-				        dev + 1, temp, gpu_degree_sign, s_fan);
-			}
-			event_abort++;
+			if (temp >= gpu_temp_limit) {
+				char s_fan[16] = "n/a";
+				if (fan >= 0)
+					sprintf(s_fan, "%u%%", fan);
+				if (!hot_gpu++ && !event_abort) {
+					log_event("Device %d overheat (%d%sC, fan %s), %s.",
+					          dev + 1, temp, gpu_degree_sign, s_fan,
+					          cool_gpu_down ? "sleeping" : "aborting job");
+					fprintf(stderr,
+					        "Device %d overheat (%d%sC, fan %s), %s.\n",
+					        dev + 1, temp, gpu_degree_sign, s_fan,
+					        cool_gpu_down ? "sleeping" : "aborting job");
+				}
+				/***
+				 * Graceful handling of GPU overheating
+				 * - sleep for a while before re-checking the temperature.
+				 ***/
+				if (cool_gpu_down) {
+					int t = 30;
+					while ((t = sleep(t)) && !event_abort);
+
+					// Re-check the temperature of the same GPU
```

Comment:
I think 30 seconds is generally too much. Since we're introducing a john.conf setting anyway, we can as well make it the number of seconds, with 0 or negative meaning to abort rather than sleep. The default can be 10 seconds. The added outer loop is probably needed in case the GPU doesn't cool down enough during the sleep, which will realistically happen e.g. if a user sets the sleep duration to 1 second.

Reply:
For the default sleep duration, we might want to tune it to maximize the average c/s rate e.g. with the limit set to 80C on Tahiti. Maybe 30 seconds is actually OK in this respect, letting the GPU continue without sleep for longer. I just don't know. I also wonder and also don't know what causes quicker wear of the GPU card (perhaps cracks in the BGA soldering, which is apparently a common problem) - frequent slight changes in temperature (e.g., 5C every 10 seconds) or less frequent but larger changes in temperature (e.g., 15C every 30 seconds).

Reply:
We already use sleep(3) for the pause file support in cracker.c, and provide it in jumbo.[ch], so perhaps the use here is OK for now. But after the next release we should probably switch to select(2)-based sleeps in all places. Code from popa3d:

```c
/*
 * A select(2)-based sleep() equivalent: no more problems with SIGALRM,
 * subsecond precision.
 */
extern int sleep_select(int sec, int usec);
```

```c
int sleep_select(int sec, int usec)
{
        struct timeval timeout;

        timeout.tv_sec = sec;
        timeout.tv_usec = usec;

        return select(0, NULL, NULL, NULL, &timeout);
}
```

We could also want to take the fcntl-based locking functions from popa3d at the same time, and stop using flock.